tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/ml/GoogleNGram.scala

```diff
@@ -0,0 +1,223 @@
+package org.allenai.nlpstack.parse.poly.ml
+
+import org.allenai.datastore._
+import org.allenai.nlpstack.core.PostaggedToken
+
+import java.io._
+
+import reming.DefaultJsonProtocol._
+
+import scala.io._
+
+/** A class that parses Google N-Gram data
+  * (http://commondatastorage.googleapis.com/books/syntactic-ngrams/index.html) to provide
+  * information about a requested n-gram.
+  * Takes the datastore location details for a data directory and parses each file, expected
+  * to be in the following format
+  * (from https://docs.google.com/document/d/14PWeoTkrnKk9H8_7CfVbdvuoFZ7jYivNTkBX2Hj7qLw/edit) -
+  *   head_word<TAB>syntactic-ngram<TAB>total_count<TAB>counts_by_year
+  * The counts_by_year format is a tab-separated list of year<comma>count items.
+  * Years are sorted in ascending order, and only years with non-zero counts are included.
+  * The syntactic-ngram format is a space-separated list of tokens, each token format is:
+  *   “word/pos-tag/dep-label/head-index”.
+  * The word field can contain any non-whitespace character.
+  * The other fields can contain any non-whitespace character except for ‘/’.
+  *   pos-tag is a Penn-Treebank part-of-speech tag.
+  *   dep-label is a stanford-basic-dependencies label.
+  *   head-index is an integer, pointing to the head of the current token.
+  *   “1” refers to the first token in the list, 2 the second,
+  *      and 0 indicates that the head is the root of the fragment.
+  */
+case class GoogleNGram(
+    groupName: String, artifactName: String, version: Int, frequencyCutoff: Int
+) {
+
+  @transient val googleNgramPath = Datastore.directoryPath(
+    groupName,
+    artifactName,
+    version
+  )
+
+  @transient val googleNgramDir = new File(googleNgramPath.toString)
+  @transient val ngramMap = GoogleNGram.constructNgramTable(googleNgramDir, frequencyCutoff)
+}
+
+/** Utility case classes to represent information associated with an ngram in the Google Ngram
+  * corpus.
+  */
+case class SyntacticInfo(word: String, posTag: String, depLabel: String, headIndex: Int)
+case class NgramInfo(syntacticNgram: Seq[SyntacticInfo], frequency: Long)
+
+/** Companion object. Contains methods to parse a Google Ngram corpus. This is not specific to
+  * the type of corpus, i.e. whether unigram, bigram, etc.
+  */
+object GoogleNGram {
```

Comment:
But keep this object name as "GoogleNGram" (just move the serializer out into a DatastoreGoogleNGram companion object).